### PR TITLE
Feature/finish proposal

### PR DESCRIPTION
**Pull request by martinyde**
https://jira.itkdev.dk/browse/FORSLAG-44

- Drush command for finishing a proposal.
- Drush command for finishing all overdue proposals.
- Cron job for finishing all overdue proposals.

### CODE REVIEW

File: web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php [hunk 1]
```diff
@@ -238,6 +240,86 @@ public function nodeEntityPresave(EntityInterface $entity): void {
     }
   }
 
+  /**
+   * Find overdue proposals.
+   *
+   * @return array|int
+   *   A list of overdue proposals
```
Comment (rimi-itk) [inline web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php:247]: Only the ids are returned, right?

File: web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php [hunk 2]
```diff
@@ -238,6 +240,86 @@ public function nodeEntityPresave(EntityInterface $entity): void {
...
+      // @todo Use method to determine support count when it is merged.
+      $supportCount = $this->connection->select('hoeringsportal_citizen_proposal_support')
+        ->condition('node_id', $entity->nid->value ?? 0)
+        ->countQuery()
+        ->execute()
+        ->fetchField();
+
+      // @todo Use constant to compare against when it is merged.
+      // Anonymize proposal if it didn't reach enough votes.
+      if ($supportCount < 10) {
```
Comment (rimi-itk) [inline web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php:296]: Consider adding a `hasEnoughVotes()`, say, function to this class to keep the code clean.

File: web/modules/custom/hoeringsportal_citizen_proposal/hoeringsportal_citizen_proposal.module [hunk 3]
```diff
@@ -61,3 +61,10 @@ function hoeringsportal_citizen_proposal_page_attachments(array &$page) {
 function hoeringsportal_citizen_proposal_node_presave(EntityInterface $entity) {
   Drupal::service(Helper::class)->nodeEntityPresave($entity);
 }
+
+/**
+ * Implements hook_cron().
+ */
+function hoeringsportal_citizen_proposal_cron() {
+  Drupal::service(Helper::class)->cron();
+}
```
Comment (rimi-itk) [inline web/modules/custom/hoeringsportal_citizen_proposal/hoeringsportal_citizen_proposal.module:70]: For consistency with similar tasks in this project, e.g. https://github.com/itk-dev/hoeringsportal/blob/develop/web/modules/custom/hoeringsportal_public_meeting/README.md, we should do this outside the Drupal cron schedule.

**Review by rimi-itk — CHANGES_REQUESTED** [review]
A comment on cron job added. The automatic code analysis has some complaints.

File: web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php [hunk 4]
```diff
@@ -255,12 +255,15 @@ public function findOverdueProposals() {
         ->condition('status', 1, '=')
         ->condition('field_content_state', 'active', '=')
         ->condition('field_vote_end', $now->format(DateTimeItemInterface::DATE_STORAGE_FORMAT), '<')
+        ->accessCheck(FALSE)
         ->execute();
```
Comment (rimi-itk) [inline web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php:259]: We should check for node type (bundle) in this query. We know that only `citizen_proposal` has the `field_content_state` and `field_vote_end` fields *for now*, but we should make sure that we don't suddenly process other node types.

File: web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php [hunk 5]
```diff
@@ -248,6 +321,20 @@ private function getProposalStorage(): PrivateTempStore {
...
+   *
+   * @param int $supportCount
+   *   The support a proposal has received.
+   *
+   * @return bool
+   *   Whether the proposal has enough votes.
+   */
+  private function hasEnoughVotes(int $supportCount): bool {
+    // @todo Use constant to compare against when it is merged.
+    return !($supportCount < 10);
```
Comment (rimi-itk) [inline web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php:335]: This has been merged now – and given rise to a merge conflict …

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: web/modules/custom/hoeringsportal_citizen_proposal/hoeringsportal_citizen_proposal.module (hunk 3 above)
Comment (martinyde) [reply web/modules/custom/hoeringsportal_citizen_proposal/hoeringsportal_citizen_proposal.module:77]: Removed and changed Readme.md

File: web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php [hunk 6]
```diff
@@ -244,9 +244,9 @@ public function nodeEntityPresave(EntityInterface $entity): void {
    * Find overdue proposals.
    *
    * @return array|int
-   *   A list of overdue proposals
+   *   A list of overdue proposal ids.
    */
-  public function findOverdueProposals() {
+  public function findOverdueProposals(): int|array {
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php]: Changed

File: web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php (hunk 4 above)
Comment (martinyde) [reply web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php:278]: Added type check

File: web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php (hunk 5 above)
Comment (martinyde) [reply web/modules/custom/hoeringsportal_citizen_proposal/src/Helper/Helper.php]: Resolved merge conflict

**Review by rimi-itk — APPROVED** [review]

**Pull request merged**